feat: git lfs pointer support extensions

zh1C opened the pull request:
`git-lfs pointer` support `--extensions` parameter to print extended information.

Resolves https://github.com/git-lfs/git-lfs/issues/4047

chrisd8088 commented:
Hey, thanks for this PR and for taking an interest in this issue!

It would indeed be a nice enhancement if the `git lfs pointer` command was aware of any configured Git LFS pointer extensions, even though these are technically still [considered](https://github.com/git-lfs/git-lfs/blob/2c7de1f90cbe13bf9c1ed43b84dda88bb32f2ba4/docs/extensions.md?plain=1#L16-L20) an experimental feature of Git LFS.

I suppose my first thought about this PR is to wonder if we want an `--extensions` command-line argument, which requires the user to (a) know about Git LFS extensions and (b) explicitly opt into checking or testing their Git LFS pointers with that extra command-line argument.

One challenge with Git LFS extensions is that some users may have them configured locally, while others do not, and since the `git lfs pointer` command is often used for low-level debugging purposes, it could be confusing if a user is trying to understand why a Git LFS pointer doesn't seem to match what their local system says it should be.

For example, suppose one person has some extensions configured and stores a `foo.bin` file in their repository in the usual way, and pushes the result, e.g.:

```
$ git config get --all --show-names --regexp 'lfs.extension.*'
lfs.extension.myext.clean my-ext-program clean -- %f
lfs.extension.myext.smudge my-ext-program smudge -- %f
lfs.extension.myext.priority 0

$ git lfs track "*.bin"
$ echo foo >foo.bin
$ git add .gitattributes foo.bin
$ git commit -m foo
$ git push origin main
```

Now someone else without those extensions configured clones the repo and examines the Git LFS pointer data to understand why they're having trouble.  The `git-lfs-pointer --file foo.bin --stdin` command is going to report `Pointers do not match`, and it does include a `ext-0-myext` line when it echoes the contents of the pointer provided on `stdin`, but it's otherwise silent as to what the difference is or what it might be caused by.

```
$ git config get --all --show-names --regexp 'lfs.extension.*'
# no output

$ git clone https://example.com/repo-with-exts
$ cd repo-with-exts

# various strange things happen, so the user tries to investigate ...

$ git show HEAD:foo.bin | git lfs pointer --file foo.bin --stdin
Git LFS pointer for foo.bin

version https://git-lfs.github.com/spec/v1
oid sha256:b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c
size 4

Git blob OID: ad483a923bcf4e7ef83182817802d740ee4d2ec6

Pointer from STDIN

version https://git-lfs.github.com/spec/v1
ext-0-myext sha256:b5bb9d8014a0f9b1d61e21e796d78dccdf1352f23cd32812f4850b878ae4944c
oid sha256:1706b5c18e4358041b463995efc30f8f721766fab0e018d50d85978b46df013c
size 4

Git blob OID: 94f284ea0ef843fccb4aba2ef5921b49995795c7

Pointers do not match
```

We can imagine other cases as well, such as if the user running `git lfs pointer` does have Git LFS extensions configured, and is trying to figure out why the ones generated by regular commands like `git add` don't match what the `git lfs pointer` command says they should be, since it's not running the raw file content through the `gitfilter.Clean()` method.  Or two users might have different Git LFS extensions installed, and so be getting different results from each other for the same initial file content.

(As a side note, these kinds of issues are, in part, why I think Git LFS pointer extensions haven't been widely adopted.)

It may require a somewhat more complex set of changes, but I think that if we're enhancing the `git lfs pointer` command to be aware of Git LFS extensions, we should offer as much guidance and help as possible for users.

For instance, the command could detect whether Git LFS extensions are configured locally, and if so, use `gitfilter.Clean()` by default when generating pointers from input data but also output a warning to `stderr` that it is doing so, with an explanation that the user could run the command again with a new `--no-extensions` option to force the command to generate a "plain" Git LFS pointer by skipping all the configured extensions and performing the direct SHA-256 hashing the command does now.

Similarly, when the command reports that two pointers do not match, if it notices that either one was generated using extensions (because some extensions are configured locally, and/or because there are `ext-<n>-<name>` lines in a raw pointer), it should include a comment in its output below the `Pointers do not match` line explaining why this might be the result of having different Git LFS extension configurations.

---

My second thought about this PR is that we always like to add tests whenever we add or change a feature, but in this case that's going to be made more difficult because we don't have a framework in our shell test suite (the one in the `t` [directory](https://github.com/git-lfs/git-lfs/tree/2c7de1f90cbe13bf9c1ed43b84dda88bb32f2ba4/t)) for testing Git LFS pointers.

Fortunately, it just so happens that I've been working on a framework to do exactly this, including a simple "case converter" Git LFS extension program and some shell test helper functions to set it up and make use of it in tests.

Unfortunately, I don't quite have that in a state ready for use yet, but it should be available soon, so if you don't mind waiting a few weeks, those test utilities should be available to be incorporated into this PR in the form of some additional tests in the `t/t-pointer.sh` test script.

---

Thank you again for getting involved with the Git LFS project!  We really appreciate contributions and new ideas like yours and they help the project develop in exciting ways.

I hope my comments above aren't discouraging; they aren't meant to be!  There's a lot of history in this project and I myself regularly find that looking into any single aspect leads to way more changes than I first anticipate.  But that's also the fun part of working on a non-trivial tool like the Git LFS client, and I hope that proves to be the case for you as well.

zh1C commented:
Hey @chrisd8088 , Thanks for the thoughtful guidance! I’ve updated this PR accordingly:

- Detect local Git LFS extensions and use `gitfilter.Clean()` by default when generating pointers.
- Added `--no-extensions` to force a plain pointer (direct SHA-256 without extensions).
- When generating with extensions, print a concise warning to stderr: “warning: Using LFS extensions, use --no-extensions for a plain pointer.”
- When pointers don’t match, if either side involves extensions (local config and/or `ext-<n>-<name>` lines), include a note: “note: Mismatch may be due to differing LFS extensions; use --no-extensions to compare plain pointers.”
- Updated the man page to document the new behavior and flag.

Happy to add tests once the new pointer/extension test utilities are available. Please let me know if you’d like any further adjustments.

chrisd8088 commented on commands/command_pointer.go:
```diff
@@ -78,18 +81,36 @@ func pointerCommand(cmd *cobra.Command, args []string) {
+
+			gitfilter := lfs.NewGitFilter(cfg)
+			hasExtensionsConfig = len(cfg.Extensions()) > 0
+			cleaned, err := gitfilter.Clean(buildFile, pointerFile, size, nil)
```
It's convenient to have a `size` value to pass here, but regarding my suggestion above to move the `io.Copy()` call into the preceding conditional block, I think we can just pass `-1` for the `fileSize` parameter and things should work OK, which means we don't need to run the contents of the pointer file through `io.Copy()` just to get the file size.

Specifically, because we're passing `nil` for the `cb` parameter of the `gitfilter.Clean()` [method](https://github.com/git-lfs/git-lfs/blob/469c8da0478f4bee02836f6dee12bce9fe737d2b/lfs/gitfilter_clean.go#L19), when the `fileSize` and `cb` parameters are [passed](https://github.com/git-lfs/git-lfs/blob/469c8da0478f4bee02836f6dee12bce9fe737d2b/lfs/gitfilter_clean.go#L52) [down](https://github.com/git-lfs/git-lfs/blob/469c8da0478f4bee02836f6dee12bce9fe737d2b/lfs/gitfilter_clean.go#L95) to the `tools.CopyWithCallback()` function, it will one [way](https://github.com/git-lfs/git-lfs/blob/469c8da0478f4bee02836f6dee12bce9fe737d2b/tools/iotools.go#L30-L32) or [another](https://github.com/git-lfs/git-lfs/blob/469c8da0478f4bee02836f6dee12bce9fe737d2b/tools/iotools.go#L30-L32) just skip using the callback, which is the only thing that really cares about knowing the file size in advance, since we otherwise just read until `EOF` is found.

(The reason the callback wants to know about the file size in advance is so that commands like `git lfs fetch` can [print](https://github.com/git-lfs/git-lfs/blob/469c8da0478f4bee02836f6dee12bce9fe737d2b/lfs/util.go#L59) intermediate progress messages like `<bytes-downloaded>/<file-size> <file-name>`.)

chrisd8088 reviewed (commented):
Hey, thanks again for working on this PR!

I've made a few initial comments, and had some minor suggestions.  Next we'll need to add some tests, of course, and I should have my pointer extension test utility ready fairly soon, so that will be useful in this PR as well.

zh1C commented on commands/command_pointer.go:
```diff
@@ -78,18 +81,36 @@ func pointerCommand(cmd *cobra.Command, args []string) {
-		fmt.Fprint(os.Stderr, tr.Tr.Get("Git LFS pointer for %s", pointerFile), "\n\n")
+		var ptr *lfs.Pointer
+		if pointerNoExtensions {
+			ptr = lfs.NewPointer(hex.EncodeToString(oidHash.Sum(nil)), size, nil)
```
Thank you for the excellent suggestion! You're absolutely right - moving the `oidHash` processing into the conditional block makes perfect sense.

chrisd8088 commented:
@zh1C—I'm very sorry it's taken so long for me to revisit this PR!

> I'm looking forward to adding the tests once your pointer extension test utility is ready.

Those tools are included in `main` now; they were developed as part of the expanded test suite for the v3.7.1 release, which we just published today.  You should be able to find examples of how to use the `lfstest-caseinverterextension` test [utility](https://github.com/git-lfs/git-lfs/blob/071e19e8ea03b1e40b181706909fb8c18d928e29/t/cmd/lfstest-caseinverterextension.go) as a pointer extension program in tests, as we've done in, for example, the new `checkout: pointer extension` [test](https://github.com/git-lfs/git-lfs/blob/071e19e8ea03b1e40b181706909fb8c18d928e29/t/t-checkout.sh#L1146-L1184) and the new `filter-process: pointer extension` [test](https://github.com/git-lfs/git-lfs/blob/071e19e8ea03b1e40b181706909fb8c18d928e29/t/t-filter-process.sh#L145-L191).

The `setup_case_inverter_extension()` [function](https://github.com/git-lfs/git-lfs/blob/071e19e8ea03b1e40b181706909fb8c18d928e29/t/testhelpers.sh#L977-L985), the `case_inverter_extension_pointer()` [function](https://github.com/git-lfs/git-lfs/blob/071e19e8ea03b1e40b181706909fb8c18d928e29/t/testhelpers.sh#L987-L991), and the `invert_case()` [function](https://github.com/git-lfs/git-lfs/blob/071e19e8ea03b1e40b181706909fb8c18d928e29/t/testhelpers.sh#L994-L996) may be useful in designing your shell tests, and the `lfstest-caseinverterextension` also writes to a log file which can be checked as well.

chrisd8088 commented on t/t-pointer.sh:
```diff
@@ -437,3 +475,224 @@ begin_test "pointer --check (with invalid arguments)"
+  
+  echo "$output" | grep "Git LFS pointer for file.dat"
+  echo "$output" | grep "ext-0-caseinverter sha256:$contents_oid"
+  echo "$output" | grep "Pointer from STDIN"
```
This all works just fine as it stands, so if you'd like to ignore my suggestion, that's perfectly OK!

Aside from dropping the `rm -f "$LFSTEST_EXT_LOG"` line, my only comment is that we can simplify these checks a bit by writing and reading files instead of using shell variables.

For instance, the following seems to work for me, and of course one can then also use `grep "Pointers do not match" pointer.log && exit 1` in the line below.

```suggestion  # Generate the pointer with extension (this is our expected format)
  git lfs pointer --file=file.dat 2>/dev/null | grep -v '\(Git LFS pointer\|warning:\)' >expected-pointer.txt

  # Test --file --stdin comparison (should match)
  cat expected-pointer.txt | git lfs pointer --file=file.dat --stdin 2>&1 | tee pointer.log

  grep "Git LFS pointer for file.dat" pointer.log
  grep "ext-0-caseinverter sha256:$contents_oid" pointer.log
  grep "Pointer from STDIN" pointer.log
```

chrisd8088 reviewed (commented):
Hey, thank you again for all the work in this PR, and I really appreciate how you've made excellent use of the `lfstest-caseinverterextension` program I wrote for the v3.7.1 release!

I've made a few final comments and suggestions on this PR, some of which are non-critical and which I'll leave to your discretion to accept or not.

There are a couple of things we do need to revise, mainly using the `... && exit 1` idiom instead of `! ...` in our shell test scripts for "negative checks" (with an extra twist when one of these is the last check in a test), and then removing some excess whitespace so our CI job's linting checks don't complain about that.

Again, thank you so much for both your contribution and your patience!  It will be great to include these enhancements to the `git lfs pointer` command in our next feature release.

chrisd8088 reviewed (approved):
Hey, thanks for accepting my various suggestions and revisions!

There's one line in one of the new tests where we need to adjust a typo, but that aside, things look great, so I'm going to provisionally approve this PR and let the CI jobs run one time to make sure they're all happy too.

chrisd8088 commented on commands/command_pointer.go:
```diff
@@ -79,17 +82,38 @@ func pointerCommand(cmd *cobra.Command, args []string) {
+			ptr = lfs.NewPointer(hex.EncodeToString(oidHash.Sum(nil)), size, nil)
+		} else {
+			setupRepository()
+
```
I'm sorry, I should have thought about this sooner, but the CI job [failure](https://github.com/git-lfs/git-lfs/actions/runs/18939990119/job/54098534817) from our "Build with latest Git" job alerted me to the implications of calling `setupRepository()` here, which [calls](https://github.com/git-lfs/git-lfs/blob/071e19e8ea03b1e40b181706909fb8c18d928e29/commands/commands.go#L315) `requireInRepo()`.

By doing so, we would inadvertently make a breaking change for anyone who is running `git lfs pointer` outside of a repository, which at the moment always succeeds.  As I've noted in a couple of recent issues (#6130, #6131), we're trying to be cautious about introducing these sorts of backwards-incompatible changes, even if they seem relatively obscure or harmless.

Instead, I'm going to recommend that we simply handle the `!cfg.InRepo()` case in the conditional block above, and remove this line:

```suggestion
```

As a sidebar, the reason the CI job fails right now is due to a combination of factors.  First, as this PR stands now, a number of the shell tests in the `t/t-pointer.sh` script which previously did not create a test repository now due so, since otherwise the tests would fail because the `requireInRepo()` function would cause the `git lfs pointer` command to exit immediately.  Second, our "Build with latest Git" job [runs](https://github.com/git-lfs/git-lfs/blob/071e19e8ea03b1e40b181706909fb8c18d928e29/.github/workflows/ci.yml#L141) the test suite with the `GIT_DEFAULT_HASH` environment variable set to `sha256`.  With that setting, when a new test repository is initialized the local Git configuration option `extensions.objectformat=sha256` is set, which then changes the OIDs returned by Git itself from 40-character SHA-1 hashes to 64-character SHA-256 hashes, and as a result, some of our existing shell tests fail because they [expect](https://github.com/git-lfs/git-lfs/blob/071e19e8ea03b1e40b181706909fb8c18d928e29/t/t-pointer.sh#L21) SHA-1 hashes from Git.

However, if we can avoid needing to modify our existing shell tests, then for the ones which don't run test repositories Git will still default to SHA-1 hashes, and we don't need to change the tests at all.  (Maybe someday Git will change their default in this case to SHA-256, but we can deal with that when it happens.)

chrisd8088 commented on commands/command_pointer.go:
```diff
@@ -79,17 +82,38 @@ func pointerCommand(cmd *cobra.Command, args []string) {
-		oidHash := sha256.New()
-		size, err := io.Copy(oidHash, buildFile)
+		var ptr *lfs.Pointer
+		if pointerNoExtensions {
```
As noted below, let's change this to the following, which allows us to leave our existing tests as-is and drop the `setupRepository()` call in the `else` block of this conditional statement.

```suggestion
		if pointerNoExtensions || !cfg.InRepo() {
```
